ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/acceptance/infrastructure/LedgerSession.scala

```diff
@@ -0,0 +1,274 @@
+
+  private lazy val commandService = CommandServiceGrpc.stub(channel)
+  private lazy val transactionService = TransactionServiceGrpc.stub(channel)
+  private lazy val ledgerIdentityService = LedgerIdentityServiceGrpc.stub(channel)
+  private lazy val partyManagementService = PartyManagementServiceGrpc.stub(channel)
+  private lazy val timeService = TimeServiceGrpc.stub(channel)
+
+  private[infrastructure] def ledgerId()(implicit context: ExecutionContext): Future[String] =
+    for {
+      response <- ledgerIdentityService.getLedgerIdentity(new GetLedgerIdentityRequest)
+    } yield response.ledgerId
+
+  private[infrastructure] def ledgerEnd()(
+      implicit context: ExecutionContext): Future[LedgerOffset] =
+    for {
+      id <- ledgerId
+      response <- transactionService.getLedgerEnd(new GetLedgerEndRequest(id))
+    } yield response.offset.get
+
+  private[infrastructure] def transactionFilter(party: String, templateIds: Seq[Identifier]) =
+    new TransactionFilter(Map(party -> filter(templateIds)))
+
+  private[infrastructure] def filter(templateIds: Seq[Identifier]): Filters =
+    new Filters(if (templateIds.isEmpty) None else Some(new InclusiveFilters(templateIds)))
+
+  private[infrastructure] def allocateParty()(implicit context: ExecutionContext): Future[String] =
+    partyManagementService.allocateParty(new AllocatePartyRequest()).map(_.partyDetails.get.party)
+
+  private[infrastructure] def allocateParties(n: Int)(
+      implicit context: ExecutionContext): Future[Vector[String]] =
+    Future.sequence(Vector.fill(n)(allocateParty()))
+
+  private[infrastructure] def transactionsUntilNow(
+      begin: LedgerOffset,
+      party: String,
+      templateIds: Identifier*)(implicit context: ExecutionContext): Future[Vector[Transaction]] =
+    for {
+      id <- ledgerId()
+      end <- ledgerEnd()
+      responses <- FiniteStreamObserver[GetTransactionsResponse](
+        transactionService
+          .getTransactions(
+            new GetTransactionsRequest(
+              ledgerId = id,
+              begin = Some(begin),
+              end = Some(end),
+              filter = Some(transactionFilter(party, templateIds)),
+              verbose = true
+            ),
+            _
+          ))
+    } yield responses.flatMap(_.transactions)
+
+  private[infrastructure] def clock()(implicit context: ExecutionContext): Future[Clock] =
+    for (id <- ledgerId()) yield new LedgerClock(id, timeService)
+
+  private[infrastructure] def create(
+      party: String,
+      templateId: Identifier,
+      args: (String, Value.Sum)*)(implicit context: ExecutionContext): Future[String] =
+    for {
+      id <- ledgerId()
+      clock <- clock()
+      now = clock.instant()
+      inFiveSeconds = now.plusSeconds(5)
+      tx <- commandService.submitAndWaitForTransaction(
+        new SubmitAndWaitRequest(
+          Some(new Commands(
+            ledgerId = id,
+            applicationId = UUID.randomUUID().toString,
```

Comment:
Keep in mind that you can only receive completions for known applicationIds.

Reply:
Each test case will have its own, automatically generated application identifier starting from https://github.com/digital-asset/daml/pull/2282/commits/d8efe8282cb58b5e7bacb4a4b2792e30b848be9e